### PR TITLE
Use Lucene index in peer recovery and resync

### PR DESCRIPTION
**Pull request by dnhatn**
We can use Lucene index exclusively in peer recoveries and primary-replica resync in 8.0.

Relates #50775

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java [hunk 1]
```diff
@@ -91,9 +90,7 @@ public void resync(final IndexShard indexShard, final ActionListener<ResyncTask>
             // Wrap translog snapshot to make it synchronized as it is accessed by different threads through SnapshotSender.
             // Even though those calls are not concurrent, snapshot.next() uses non-synchronized state and is not multi-thread-compatible
             // Also fail the resync early if the shard is shutting down
-            snapshot = indexShard.getHistoryOperations("resync",
-                indexShard.indexSettings.isSoftDeleteEnabled() ? Engine.HistorySource.INDEX : Engine.HistorySource.TRANSLOG,
-                startingSeqNo);
+            snapshot = indexShard.newChangesSnapshot("resync", startingSeqNo, Long.MAX_VALUE, false);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java]: `readHistoryOperations` used `Math.max(0, startingSeqNo)` as `from`. I wonder if we need to guard against `SequenceNumbers.UNASSIGNED_SEQ_NO` similarly here.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 2]
```diff
@@ -296,9 +286,8 @@ public void recoverToTarget(ActionListener<RecoveryResponse> listener) {
                     shardId + " initiating tracking of " + request.targetAllocationId(), shard, cancellableThreads, logger);
 
                 final long endingSeqNo = shard.seqNoStats().getMaxSeqNo();
-                logger.trace("snapshot translog for recovery; current size is [{}]",
-                    shard.estimateNumberOfHistoryOperations("peer-recovery", historySource, startingSeqNo));
-                final Translog.Snapshot phase2Snapshot = shard.getHistoryOperations("peer-recovery", historySource, startingSeqNo);
+                logger.trace("snapshot for recovery; current size is [{}]", estimateNumberOfHistoryOperations(startingSeqNo));
+                final Translog.Snapshot phase2Snapshot = shard.newChangesSnapshot("peer-recovery", startingSeqNo, Long.MAX_VALUE, false);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: can `startingSeqNo` be `UNASSIGNED_SEQ_NO here?

**Review by ywelsch — COMMENTED** [review]
I've left one comment, looking good o.w.

File: server/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java]: I could not find the reason why we used `Math.max(0, startingSeqNo)`. We should always have a valid global checkpoint here, then startingSeqNo should be non-negative. I added an assertion in https://github.com/elastic/elasticsearch/pull/51189/commits/ea8da99f2577a470471f6c2dfc94cb9ccb46b72d.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java (hunk 2 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: It should be non-negative. We assert it here: https://github.com/elastic/elasticsearch/blob/106b0d7789e150bda8077105e751033453c54b16/server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java#L269.

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by dnhatn** [comment]
Thanks Yannick

**Pull request merged**